spring-integration-core/src/main/java/org/springframework/integration/support/json/EmbeddedJsonHeadersMessageMapper.java

```diff
@@ -0,0 +1,255 @@
+ * <p>
+ * <b>IMPORTANT</b>
+ * <p>
+ * The default object mapper uses default typing. From the Jacskson Javadocs:
+ * <p>
+ * NOTE: use of Default Typing can be a potential security risk if incoming content comes
+ * from untrusted sources, and it is recommended that this is either not done, or, if
+ * enabled, use {@link ObjectMapper#setDefaultTyping} passing a custom
+ * {@link com.fasterxml.jackson.databind.jsontype.TypeResolverBuilder} implementation that
+ * white-lists legal types to use.
+ * <p>
+ * A constructor is provided allowing the provision of such a configured object mapper.
+ *
+ * @author Gary Russell
+ * @since 5.0
+ *
+ */
+public class EmbeddedJsonHeadersMessageMapper implements BytesMessageMapper {
+
+	private final ObjectMapper objectMapper;
+
+	private final List<Pattern> headerPatterns;
+
+	private final boolean allHeaders;
+
+	private boolean rawBytes = true;
+
+	/**
+	 * Construct an instance that embeds all headers, using the default
+	 * JSON Object mapper.
+	 */
+	public EmbeddedJsonHeadersMessageMapper() {
+		this(Collections.singleton(Pattern.compile(".*")));
+	}
+
+	/**
+	 * Construct an instance that embeds headers matching the supplied patterns, using
+	 * the default JSON object mapper.
+	 * @param headerPatterns the patterns.
+	 */
+	public EmbeddedJsonHeadersMessageMapper(Collection<Pattern> headerPatterns) {
+		this(JacksonJsonUtils.messagingAwareMapper(), headerPatterns);
+	}
+
+	/**
+	 * Construct an instance that embeds headers matching the supplied patterns using the
+	 * supplied JSON object mapper.
+	 * @param objectMapper the object mapper.
+	 * @param headerPatterns the patterns.
+	 */
+	public EmbeddedJsonHeadersMessageMapper(ObjectMapper objectMapper, Collection<Pattern> headerPatterns) {
+		this.objectMapper = objectMapper;
+		this.headerPatterns = new ArrayList<>(headerPatterns);
+		this.allHeaders = this.headerPatterns.size() == 1 && this.headerPatterns.get(0).pattern().equals(".*");
+	}
+
+	/**
+	 * Set to false to encode byte[] payloads as base64 in JSON. When true,
+	 * byte[] payloads are rendered as discussed in the class description.
+	 * @param rawBytes false to encode as base64.
+	 */
+	public void setRawBytes(boolean rawBytes) {
+		this.rawBytes = rawBytes;
+	}
+
+	public Collection<Pattern> getHeaderPatterns() {
+		return Collections.unmodifiableCollection(this.headerPatterns);
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public byte[] fromMessage(Message<?> message) throws Exception {
+		Message<?> messageToEncode = this.allHeaders ? message : pruneHeaders(message);
+		if (this.rawBytes && message.getPayload() instanceof byte[]) {
+			return fromBytesPayload((Message<byte[]>) messageToEncode);
+		}
+		else {
+			return this.objectMapper.writeValueAsBytes(messageToEncode);
+		}
+	}
+
+	private Message<?> pruneHeaders(Message<?> message) {
+		Map<String, Object> headersToEmbed = new HashMap<>();
+		message.getHeaders().forEach((k, v) -> {
```

Comment:
````java
Stream<Pattern> patternStream = this.headerPatterns.stream();
		
		Map<String, Object> headersToEmbed =
				message.getHeaders()
						.entrySet()
						.stream()
						.filter(e -> patternStream.anyMatch(p -> p.matcher(e.getKey()).matches()))
						.collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
````
?

Also I would say the `this.headerPatterns.stream()` should be as a property, too. To avoid extra `Stream` object for each message.

Reply:
You can't reuse the patternStream like that; it will already be partially (or totally) consumed.